### PR TITLE
s4698277 - 2D UNet Segmentation

### PR DESCRIPTION
**Pull request by Woahseph**
Hey Shakes, this pull request implements a 2D UNet algorithm on the HipMRI study on prostate data using PyTorch. This includes loading the Nifti files into numpy arrays, then training and testing the model on this data. The evaluation metric of the model is based on the dice similarity coefficients, which should be greater than 0.75 for all classes, but primarily the prostate class as this is the significant region of interest. Unfortunately, this model had quite a low dice score for the prostate class. However, several of the predictions were fairly accurate, and it is thought that the dice score is calculated incorrectly for when the class is not in the label. Please see `README.md` for more details.

The following files are being submitted:

`dataset.py`: For loading Nifti images into PyTorch datasets.

`modules.py`: Contains the 2D UNet model architecture.

`predict.py`: Loads a saved model to predict a randomly sampled image from the validation dataset. The predicted mask is plotted against the true mask to visually identify the performance of the model. The dice scores are also computed for further analysis and understanding.

`train.py`: Trains the 2D UNet model on training data and saves the completed model with plots of the training metrics. The model is then used on the test data to evaluate the performance based on the class dice scores, primarily the prostate class.

`utils.py`: Contains helper functions used to load the Nifti files into numpy arrays.

`README.md`: Detailed documentation that expands upon task and model descriptions, the model training and testing results, and finally the usage instructions.

As briefly mentioned above, further explanations are available in `README.md`. Any feedback on this pull request is appreciated and will be incorporated! Please let me know also if there are any issues or concerns.

### CODE REVIEW

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Half of labels reached DICE over 0.75 on testing set, but as shown in dice plot of training, training dice is around 0.2. Listed some good and bad examples from testing, dice varied a lot. Training loss was provided. Nice documentation. Missed reference. 
2. **Implementation**: Nice uni tests. Some hyperparameters were hard-coded. Overall, nice and clear.
3. **Design**: Good
4. **Comment**: Nice and sufficient, detailed and well-structured. 
5. **Difficulty**: Easy

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally
